### PR TITLE
PartSysBase and SimpleSOL+particles tweaks

### PR DESCRIPTION
**Pull request by oparry-ukaea**
# Description
Move handling of particle output frequency into particle system base class.

## Type of change

- [x] Refactoring/cleanup
~- [ ] Bug fix (non-breaking change)~
~- [ ] New feature (non-breaking change which adds functionality)~
~- [ ] API change (breaks downstream workflows and requires major semver bump)~
~- [ ] Requires documentation updates~

# Testing

No new tests. Passes existing unit and integration tests.

Test Configuration:

OS: Ubuntu 22.04
Compiler: GCC 11.3.0 / OneAPI v2022.1.0
SYCL implementation: Hipsycl v0.9.2 / DPC++ v2022.1.0
MPI details: MPICH v4.0.2
Hardware: CPU (Intel Alder Lake)

# Checklist:

- [x] New and existing unit tests pass locally with my changes
- [x] I have used understandable variable names
- [x] I have run `clang-format` against my `*.hpp` and `*.cpp` changes
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
~- [ ] I have added tests that prove my fix is effective or that my feature works~
~- [ ] Any new dependencies are automatically built for users via `cmake`~
~- [ ] I have run `cmake-format` against my changes to `CMakeLists.txt`~
~- [ ] I have run `black` against changes to `*.py`~
~- [ ] I have made corresponding changes to the documentation~
```

### CODE REVIEW

**Comment by oparry-ukaea** [comment]
N.B. I've only changed the parameter name in the SimpleSOL config file for now, since that's the only solver using PartSysBase at the moment.
I'll update the HW/LAPD solvers and their test/example configs on feature/3DHW.

**Comment by oparry-ukaea** [comment]
Last commit just adds some explanation of how 'num_particles_total' is used in SimpleSOL.

File: include/nektar_interface/solver_base/partsys_base.hpp [hunk 1]
```diff
@@ -32,98 +32,43 @@ class PartSysBase {
...
+
+  /// @brief Clear up memory related to the particle system
+  inline void free();
 
   /**
-   * @brief Clear up memory related to the particle system
+   * @brief Return true if \p step is a scheduled output step, according to the
+   * frequency read from the config file, false otherwise
+   *
+   * @ param step
```
Comment (will-saunders-ukaea) [inline include/nektar_interface/solver_base/partsys_base.hpp:49]: Maybe use the `@returns`

File: include/nektar_interface/solver_base/partsys_base.hpp [hunk 2]
```diff
@@ -150,92 +95,174 @@ class PartSysBase {
...
-               "' or "
-               "'" +
-               NUM_PARTS_PER_CELL_STR + "' are set)");
-      }
-    }
-    report_param("Total number of particles", this->num_parts_tot);
+  std::map<std::string, std::string> param_vals_to_report;
+};
+
+inline PartSysBase::PartSysBase(const LU::SessionReaderSharedPtr session,
```
Comment (will-saunders-ukaea) [inline include/nektar_interface/solver_base/partsys_base.hpp:126]: If the definitions are split from the implementations is it worth putting them in a source cpp file in `src/nektar_interface/solver_base/partsys_base.cpp`?

Comment (will-saunders-ukaea) [reply include/nektar_interface/solver_base/partsys_base.hpp:126]: It would mean all the expensive to compile templated loops are done once.

Comment (oparry-ukaea) [reply include/nektar_interface/solver_base/partsys_base.hpp:126]: Good point, I was confusing NESOlib with NP and forgetting it wasn't header-only.  I'll split it up.

Comment (oparry-ukaea) [reply include/nektar_interface/solver_base/partsys_base.hpp]: Done

File: include/nektar_interface/solver_base/partsys_base.hpp (hunk 1 above)
Comment (oparry-ukaea) [reply include/nektar_interface/solver_base/partsys_base.hpp]: Done

**Comment by will-saunders-ukaea** [comment]
Unit+Integration tests passed with hipsycl.

**Review by will-saunders-ukaea — APPROVED** [review]

**Pull request merged**